### PR TITLE
Interface model: IModel first implementation

### LINKED ISSUE #9: Enable support of simultaneous loading of multiple models
Currently, only one IFC model and/or one point cloud model can be loaded at once. Enable simultaneous loading and viewing of multiple files.

Discuss the functionality with @pjanck and/or @hechth.

### LINKED ISSUE #108: Order in loading files
When simultaneously rendering a point cloud and an ifc model, the point cloud should be loaded first. If ifc is loaded first and the point cloud afterwards, the ifc model disappears.

### LINKED ISSUE #149: [BUG] Application crashes when loading two IFC files sequentially
**Describe the bug**
The application crashes when two IFC files are loaded after another.
Tested with example files [surface-model.ifc](https://github.com/tumcms/Open-Infra-Platform/blob/development/testdata/IFC4x1ExampleFiles/official/E2_Basic_geometric_shape/surface-model.ifc) and [tessellated-item.ifc](https://github.com/tumcms/Open-Infra-Platform/blob/development/testdata/IFC4x1ExampleFiles/official/E2_Basic_geometric_shape/tessellated-item.ifc).

**To Reproduce**
1. Start application.
2. Open first file.
3. Open second file.
4. See error [log.txt](https://github.com/tumcms/Open-Infra-Platform/files/4680131/log.txt)

**Expected behavior**
Both files should open normally after another.

**CMake configuration**
 If applicable, list any special configurations you have made in CMake.
- EARLYBINDING_WITH_IFC4x1: selected

**Additional context**
See [log.txt](https://github.com/tumcms/Open-Infra-Platform/files/4680131/log.txt)

### PR DESCRIPTION
**Pull request by pjanck**
First implementation is there. `IModel` knows about its:
- extent (`oip::BBox`) - needed in Viewport (implemented)
- EPSG code (`std::string`) - needed by me for my PhD (not implemented yet)

Still missing:
- `data` would need an array of `IModel`s which get filled when reading files (currently it is left as it is)
- model deletion/addition should be considered
- UI overview of loaded models (with zoom-to feature)

*edit 20201025*: Fixes #9 closes #108 resolves #149

### CODE REVIEW

File: Base/src/IModel.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	#pragma region Interface georeferencing
+
+    // needed for correct correlation between multiple models
+    // also to determine if a transformation would be needed
+	// the function should return empty string if unknown
+    virtual std::string getEPSGcode() const = 0;
+
+	// change the georeferencing coordinate reference system to the newEPSGcode
+	// function takes three coordinate values and returns three transformed coordinates
+    virtual void transformAllPoints(const std::string& newEPSGcode, std::function<std::tuple<double,double,double> const(double,double,double)>& transf) = 0;
```
Comment (hechth) [inline Base/src/IModel.h]: Why 3 doubles? I OIP should work with it's own data format for 3d points I assume!

Maybe also too specific for the general model.

File: Base/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#    Qt5::Xml
+#    Qt5::XmlPatterns
+    BlueFramework.Core
+    BlueFramework.Rasterizer
+#    BlueFramework.Application
+    carve
+#    eigen
+)
+
+#set_target_properties(OpenInfraPlatform.Base 	PROPERTIES FOLDER "OpenInfraPlatform")
```
Comment (hechth) [inline Base/CMakeLists.txt]: Remove out-commented stuff if checking in new file!

File: Base/src/IModel.h [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    virtual oip::BBox     getExtent() = 0;
+
+	#pragma endregion Interface bounding box
+
+	#pragma region Interface georeferencing
+
+    // needed for correct correlation between multiple models
+    // also to determine if a transformation would be needed
+	// the function should return empty string if unknown
+    virtual std::string getEPSGcode() const = 0;
```
Comment (hechth) [inline Base/src/IModel.h]: Maybe move this to another interface called IGeoReferencable?

Comment (pjanck) [reply Base/src/IModel.h]: I can see the idea, however, every model should be georeferencable. 

Proposal: Make a new interface `IGeoReferencable`, that the `IModel` includes, but not implements?

File: Base/CMakeLists.txt (hunk 2 above)
Comment (pjanck) [reply Base/CMakeLists.txt]: Right. Resolved 30dedf2

File: Base/src/IModel.h (hunk 1 above)
Comment (pjanck) [reply Base/src/IModel.h]: Well - 3d coordinates are transformed to 3d coordinates. I am open to suggestions, what to include here :)

**Comment by pjanck** [comment]
Fixes #9 .

Implemented:
- `data` has an array of models, with getters/setters
- `GeorefMetadata` introduced to handle georeferencing metadata

Still missing:
- [x] dialog/UI element with an overview over the loaded models
- [ ] zoom-to feature (see #228 )
- [ ] loading IFC files of different IFC version (see #213 )

**Comment by pjanck** [comment]
View after the last commit, with these files loaded (and zoomed in a bit):
- tessellated-item.ifc
- basin-faceted-brep.ifc

grafik

**Comment by pjanck** [comment]
Upgrade: the UI now shows some moels' properties on the right (same models loaded as above).

grafik

**Comment by pjanck** [comment]
With the latest changes, the offsets are now calculated correctly and models rendered in correct spatial relationship to one another.

grafik

**Pull request merged**